lexicon/bpe.py

```diff
@@ -0,0 +1,141 @@
+        :param bpe_vocab: vocab file to limit which bpe splits can be created
+        :param subword_nmt_repo: cloned repository
+        :param unk_label: unknown label, used in case a BPE token is created that is not in the vocab.
+        :param vocab_blacklist: which bpe_vocab entries not to load into the "phoneme/bpe-token" inventory
+            e.g. remove "<s>" and "</s>"
+        :param keep_special_lemmas: If special lemmas should be kept,
+            usually yes for RASR search and no for Flashlight search.
+            The phonemes of the special lemmas will also be kept, therefore
+            make sure there is no overlap with the BPE vocab.
+        """
+        self.base_lexicon_path = base_lexicon_path
+        self.bpe_codes = bpe_codes
+        self.bpe_vocab = bpe_vocab
+        self.subword_nmt_repo = subword_nmt_repo
+        self.unk_label = unk_label
+        if vocab_blacklist is None:
+            self.vocab_blacklist = set()
+        else:
+            # convert list to set for faster "in" check
+            self.vocab_blacklist = set(vocab_blacklist)
+        self.keep_special_lemmas = keep_special_lemmas
+
+        self.out_lexicon = self.output_path("lexicon.xml.gz", cached=True)
+
+    def tasks(self):
+        yield Task("run", resume="run", mini_task=True)
+
+    def _fill_lm_tokens(self, base_lexicon: Lexicon):
+        lm_tokens = set()
+        special_lemmas = []
+        for lemma in base_lexicon.lemmata:
+            if lemma.special is None:
+                for orth in lemma.orth:
+                    lm_tokens.add(orth)
+            else:
+                special_lemmas.append(lemma)
+
+        return sorted(lm_tokens), special_lemmas
+
+    def _fill_vocab_and_lexicon(self):
+        lexicon = Lexicon()
+        vocab = set()
+        with util.uopen(self.bpe_vocab.get_path(), "rt") as f, util.uopen("fake_count_vocab.txt", "wt") as vocab_file:
+            for line in f:
+                line = line.strip()
+                if line == "{" or line == "}":
+                    continue
+                symbol = line.split(":")[0][1:-1]
+                if symbol not in self.vocab_blacklist:
+                    # Fake count vocab filled with -1 so that all merges possible are done
+                    vocab_file.write(symbol + " -1\n")
+                    symbol = symbol.replace(".", "_")
+                    vocab.add(symbol)
+                    lexicon.add_phoneme(symbol.replace(".", "_"))
+
+        return vocab, lexicon
+
+    def run(self):
+        base_lexicon = Lexicon()
+        base_lexicon.load(self.base_lexicon_path)
+
+        lm_tokens, special_lemmas = self._fill_lm_tokens(base_lexicon)
+
+        with util.uopen("words", "wt") as f:
+            for t in lm_tokens:
+                f.write(f"{t}\n")
+
+        vocab, lexicon = self._fill_vocab_and_lexicon()
+
+        # add special lemmas back to lexicon
+        if self.keep_special_lemmas is True:
```

Comment:
```suggestion
        if self.keep_special_lemmas:
```

Reply:
I would keep the `is True` check, otherwise other objects might evaluate this to true and I don't think we should allow this.